server/main_server/src/main.cpp

```diff
@@ -0,0 +1,1342 @@
+        // This process should never introduce threads. Everything it cares about must be handled
+        // within the loop. This keeps things simple and straight forward.
+        //
+        // THE PARENT PROCESS IS THE ONLY PROCESS THAT SHOULD/CAN REACT TO SIGNALS!
+        // EVERYTHING IS PROPAGATED THROUGH/FROM THE PARENT PROCESS!
+
+        // dsm = Short for delay server migration
+        // This is used to control the frequency of the delay server migration logic.
+        auto dsm_time_start = std::chrono::steady_clock::now();
+
+        // stfp = Short for stacktrace file processor
+        // This is used to control the frequency of the stacktrace file processing logic. 
+        auto stfp_time_start = dsm_time_start;
+
+        while (true) {
+            if (g_terminate) {
+                log_server::info("{}: Received shutdown instruction. Exiting server main loop.", __func__);
+                handle_shutdown();
+                break;
+            }
+
+            if (g_terminate_graceful) {
+                log_server::info("{}: Received graceful shutdown instruction. Exiting server main loop.", __func__);
+                handle_shutdown_graceful();
+                break;
+            }
+
+            if (g_reload_config) {
+                handle_configuration_reload(write_to_stdout, enable_test_mode);
+            }
+
+            // Clean up any zombie child processes if they exist. These appear following a configuration
+            // reload. We call waitpid() multiple times because the main server processes may have multiple
+            // child processes.
+            for (int i = 0; i < g_max_number_of_child_processes; ++i) {
+                waitpid(-1, nullptr, WNOHANG);
+            }
+
+            log_stacktrace_files(stfp_time_start);
+            remove_leftover_agent_info_files_for_ips();
+            launch_agent_factory(__func__, write_to_stdout, enable_test_mode);
+            migrate_and_launch_delay_server(write_to_stdout, enable_test_mode, dsm_time_start);
+
+            std::this_thread::sleep_for(std::chrono::seconds{1});
+        }
+
+        log_server::info("{}: Server shutdown complete.", __func__);
+
+        return 0;
+    }
+    catch (const std::exception& e) {
+        fmt::print(stderr, "Error: {}\n", e.what());
+        return 1;
+    }
+} // main
+
+namespace
+{
+    auto print_usage() -> void
+    {
+        // TODO Update help text.
+        fmt::print(
+R"__(irodsServer - Launch an iRODS server
+
+Usage: irodsServer [OPTION]...
+
+TODO More words ...
+
+Mandatory arguments to long options are mandatory for short options too.
+
+Options:
+  -d, --daemonize
+                Run server instance in the background as a service.
+  -p, --pid-file FILE
+                The absolute path to FILE, which will be used as the
+                PID file for the server instance.
+  -h, --help    Display this help message and exit.
+  -v, --version Display version information and exit.
+)__");
```

Comment:
Handle TODO.

Reply:
Also, need to do the following:
- Add `--stdout`
- Add `--test-mode`
- Do not allow `-d` and `--stdout` to be used at the same time

Reply:
Updated help text.

- Added description
- Adjusted wording for options
- Adjusted order of options
- Added section about environment variables
- Added section about signals